sources/api/bootstrap-commands/src/main.rs

```diff
@@ -0,0 +1,315 @@
+use snafu::{ensure, OptionExt, ResultExt};
+use std::collections::BTreeMap;
+
+use std::env;
+use std::ffi::OsStr;
+use std::fs;
+use std::path::{Path, PathBuf};
+use std::process::{self, Command};
+use std::str::FromStr;
+
+use bottlerocket_modeled_types::{BootstrapMode, Identifier};
+use bottlerocket_settings_models::ValidCommand;
+
+const DEFAULT_CONFIG_PATH: &str = "/etc/bootstrap-commands/bootstrap-commands.toml";
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+struct BootstrapCommandConfig {
+    bootstrap_commands: Option<BTreeMap<Identifier, BootstrapCommand>>,
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+struct BootstrapCommand {
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    commands: Option<Vec<ValidCommand>>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    mode: Option<BootstrapMode>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    essential: Option<bool>,
+}
+
+/// Stores user-supplied global arguments
+struct Args {
+    log_level: LevelFilter,
+    config_path: PathBuf,
+}
+
+impl Default for Args {
+    fn default() -> Self {
+        Self {
+            log_level: LevelFilter::Info,
+            config_path: PathBuf::from_str(DEFAULT_CONFIG_PATH).unwrap(),
+        }
+    }
+}
+
+/// Wrapper around process::Command that adds error checking.
+fn run_command<I, S>(bin_path: &str, args: I) -> Result<()>
+where
+    I: IntoIterator<Item = S>,
+    S: AsRef<OsStr>,
+{
+    let mut command = Command::new(bin_path);
+
+    command
+        .args(args)
+        .status()
+        .context(error::ExecutionFailureSnafu { command })?;
+
+    Ok(())
+}
+
+/// Handles running of the api commands in a Bootstrap command
+fn handle_bootstrap_command<S>(name: S, bootstrap_command: &BootstrapCommand) -> Result<()>
+where
+    S: AsRef<str>,
+{
+    let name = name.as_ref();
+    let mode = bootstrap_command.mode.clone().unwrap_or_default();
+    let commands = bootstrap_command.commands.clone().unwrap_or_default();
+
+    info!("Running bootstrap command with name: {}", name);
+
+    if mode == "off" {
+        // If mode is 'off', just log this information.
+        info!("Bootstrap command mode for '{}' is 'off'", name);
+    } else {
+        info!("Processing Bootstrap command '{}' ... ", name);
+
+        for command in commands.iter() {
+            // this condition is already taken care of
+            let (cmd, args) = command.get_args();
+            let arg_str: Vec<&str> = args.iter().map(String::as_str).collect();
+            run_command(cmd.as_str(), arg_str)?;
+        }
+
+        if mode == "once" {
+            let formatted = format!("settings.bootstrap-commands.{}.mode=off", name);
+            info!("Turning off bootstrap-command '{}'", name);
+            run_command("apiclient", ["set", formatted.as_str()])?;
+        }
```

Comment:
The problem with this approach is that if one of the commands is `apiclient reboot`, then we may not finish changing the mode before the reboot kicks in. Same (but more subtly) if the command is `apiclient update apply -c -r` for a one-time update at boot.

To fix this, you can do something like:
1. Check for `/var/lib/bottlerocket/bootstrap-commands@<name>.ran` prior to executing a "once" command. If found, skip the command.
2. Create `/var/lib/bottlerocket/bootstrap-commands@<name>.ran` prior to executing a "once" command. This records that the command has been initiated.
3. For any "off" command, remove the marker file if it exists, as the user might later re-enable it as a "once" command and expect it to run again.

This will satisfy the first part of the contract - making sure a command only runs once.

For the second part, updating the API settings to "off" to reflect that the commands have run, you could maintain a set of bootstrap commands to disable, and disable them all at once prior to returning a value from the overall program execution.

Reply:
We opted on solving this with systemd-inhibit.